### PR TITLE
ipv6/rpl: add hop-by-hop extension handling

### PR DESCRIPTION
**Pull request by BytesGalore**
This PR introduces hop-by-hop header handling for RPL, 
which includes an adjustment to extension header processing in `gnrc_ipv6`.

For receiving packets the adjustment introduces a sequential procedure which enables all interested threads to evaluate an extension and react before the packet is forwarded further.
When sending a packet all interested threads will be called to insert their extension before the packet is sent.

### CODE REVIEW

File: sys/include/net/gnrc/rpl/hop.h [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+ *          RFC 6553
+ *      </a>
+ */
+typedef struct __attribute__((packed)) {
+    uint8_t nh;           /**< next header identifier (0x63) */
+    uint8_t len;          /**< length in 8 octets without first octet */
+    uint8_t ORF_flags;    /**< O|R|F flags followed by 5 unused bits */
+    uint8_t instance_id;  /**< id of the instance */
+    uint16_t sender_rank; /**< rank of the sender */
+} gnrc_rpl_hop_opt_t;
```
Comment (miri64) [inline sys/include/net/gnrc/rpl/hop.h]: Everything above looks more like RPL-related than GNRC related stuff. Why not move it to `net/rpl[/hop].h`?

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c [hunk 2]
```diff
@@ -97,6 +97,49 @@ kernel_pid_t gnrc_ipv6_init(void)
...
+                          , PROTNUM_IPV6_EXT_FRAG
+                          , PROTNUM_IPV6_EXT_ESP
+                          , PROTNUM_IPV6_EXT_AH
+                          , PROTNUM_IPV6_EXT_DST
+                          , PROTNUM_IPV6_EXT_MOB
+                          };
+    bool encapsulate = false;
+
+    for ( size_t i = 0; i < sizeof(ext_demux); ++i ) {
+        gnrc_netreg_entry_t *call = gnrc_netreg_lookup(GNRC_NETAPI_MSG_TYPE_SND, ext_demux[i]);
```
Comment (miri64) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: `GNRC_NETAPI_MSG_TYPE_SND` isn't a `NETTYPE`

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c [hunk 3]
```diff
@@ -97,6 +97,49 @@ kernel_pid_t gnrc_ipv6_init(void)
...
+        msg_t msg, rcv;
+        msg.type = ext_demux[i];
+
+        while (call) {
+            if (   ext_demux[i] == PROTNUM_IPV6_EXT_HOPOPT
+                || ext_demux[i] == PROTNUM_IPV6_EXT_RH ) {
+                encapsulate = true;
+            }
+            msg.content.ptr = pkt;
+            msg_send_receive(&msg, &rcv, call->target.pid);
```
Comment (miri64) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: I'm not sure if you use `NETAPI_SND` here, because above you use `NETAPI_SND` in a weird context, but `NETAPI_SND` uses asynchronous API, so this messes up things.

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c [hunk 4]
```diff
@@ -97,6 +97,49 @@ kernel_pid_t gnrc_ipv6_init(void)
...
+        }
+    }
+
+    if (encapsulate) {
+        ipv6_hdr_t* hdr = gnrc_ipv6_get_header(pkt);
+        if (hdr) {
+            gnrc_pktsnip_t* tmp = pkt;
+            pkt = gnrc_ipv6_hdr_build(pkt, &hdr->src, &hdr->dst);
+            if (pkt == NULL) {
+                /* revert but what do we now */
```
Comment (miri64) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: abort sending, I would say.

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c [hunk 5]
```diff
@@ -285,7 +329,18 @@ static void *_event_loop(void *args)
                 reply.content.value = -ENOTSUP;
                 msg_reply(&msg, &reply);
                 break;
-
+            case GNRC_IPV6_EXT_HANDLE_NEXT_HDR:
```
Comment (miri64) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: Why not do all below in `gnrc_ipv6_ext_demux()`?

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c [hunk 6]
```diff
@@ -919,49 +974,11 @@ static void _receive(gnrc_pktsnip_t *pkt)
...
-                }
-                next = ptr->next;
-                ptr->next = reversed_pkt;
-                reversed_pkt = ptr;
-                ptr = next;
-            }
-            _send(reversed_pkt, false);
-            return;
-        }
-        else {
```
Comment (miri64) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: WTF?

**Comment by miri64** [comment]
This whole PR seems very intrusive into `gnrc_ipv6` just to handle hop-by-hop headers and additionally seems to strongly misuse GNRC's API for things that can be done in a more straight forward way.

File: sys/include/net/gnrc/rpl/hop.h (hunk 1 above)
Comment (BytesGalore) [inline sys/include/net/gnrc/rpl/hop.h]: I've put it alongside the `gnrc/rpl/srh` extension handler, to group the extensions.
Migrating both to `net/rpl` could be the way.

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 2 above)
Comment (BytesGalore) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: Hm, seems I didn't pushed it. It should be `GNRC_NETTYPE_IPV6`

Comment (miri64) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: Even then this doesn't seem to make sense, since you are not sending here but building headers. Why not just use functions or if you absolutely have to use IPC (which I think would slow-down the sending process majorly), use raw IPC?

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 3 above)
Comment (BytesGalore) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: The idea here is to consecutively send a **blocking** message to allow each interested thread to add their extension header to the IPv6 packet and inform the IPv6 thread back when its done.
This way all extension headers can be inserted in a single packet.

Trying this asynchronously would result in sending a multitude of copies of the packet, since all threads do a copy-on-write operation to the packet when inserting an extension header.

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 4 above)
Comment (BytesGalore) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: yup, makes sense.

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 3 above)
Comment (miri64) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: That's why I stated below, that this is a misuse of `NETAPI_SND`. You are not sending here, you are building headers, so I think a callback function to the header builders would be more sensible here.

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 5 above)
Comment (BytesGalore) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: This is a direct reply of a thread using plain msg passing when parsing an extension from a received IPv6 packet is finished.
The `*handle` is used as an iteration message by the thread to point to the next header to be processed, which is eventually dispatched in `gnrc_ipv6_ext_demux()`.
The IPv6 thread needs to wait forwarding the IPv6 packet further until all extensions has been processed by all interested threads.

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 6 above)
Comment (BytesGalore) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: We must prevent to forward the packet before all extension headers have been processed.
Currently I moved the responsibility for preparing the packet as above to the thread that processed the last extension header.
But as you stated, the IPv6 thread is the responsible one for doing it and we should keep it that way.

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 2 above)
Comment (BytesGalore) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: I don't like the fact that IPv6 needs to know something about, say RPL.
Additionally it would require some handshaking which is not needed by using IPC. 
> use raw IPC

I do, I iterate all interested PIDs with `gnrc_netreg_lookup()` and use plain message passing here.

**Comment by BytesGalore** [comment]
>  just to handle hop-by-hop headers

I disagree, 
it allows to process any extension header by other threads and it uses the already present information and mechanics form `netreg` for interest registration on extension header and their dispatching.
It is not bound to RPL or Hop-by-Hop extensions and such mechanism is required for any cross-layer extension header handling, e.g. RPL source routing headers.

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 5 above)
Comment (miri64) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: Ok, but in this case please surround the case with a suitable `#ifdef MODULE_...`

**Comment by miri64** [comment]
> I disagree,
it allows to process any extension header by other threads and it uses the already present information and mechanics form netreg for interest registration on extension header and their dispatching.
It is not bound to RPL or Hop-by-Hop extensions and such mechanism is required for any cross-layer extension header handling, e.g. RPL source routing headers.

But it changes the semantics of that mechanics and their API significantly, which can have unforeseen side-effects and makes potential refactoring work harder. I also don't think that handling *IPv6* extension headers in any other thread than the IPv6 handling thread is a good idea. Every synchronous context switch demands time and message queue space (thus blocking the queue for other packages). I'm actually currently trying to reduce those within the stack. If you absolutely want to use `netreg` for this (but maybe use `NETAPI_GET` for it with a new `NETOPT_EXT_HDR` option) consider using the [`netreg` callback type](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/net/gnrc/netreg.h#L51) (and extending it for get/set, as it currently only works with snd/rcv) instead of the default type.

**Comment by BytesGalore** [comment]
> But it changes the semantics of that mechanics and their API significantly

I don't see your point.
The changes introduced by this PR extend the eventloop to recognize one new message type and doesn't change anything in the API semantics.
It introduces a sequentially ordered processing of extension headers in registered contexts, in which demuxing an extension header passes the responsibility to the informed thread which in turn passes pktsnip and the responsibility back here when it finished processing the header.

Using the `netreg` callback module as you proposed is an interesting feature, but it would bypass the regular processing order of the called context which could result in unforeseen states.

Could you rephrase your proposition for using `NETAPI_GET/SET`.
If I understand right, you would propose to use it to replace the direct IPC calls, right?

**Comment by miri64** [comment]
I was talking about [this function](https://github.com/RIOT-OS/RIOT/pull/7231/files#diff-74976d45a42eaa063e3561b5fd36f615R100), that used to have a `NETAPI_SND` in it to make synchronous (`msg_send_receive()`) IPC calls. I now see that this changed and you use the `EXT_TYPE` as message type. I still think however, that it might be more elegant to use `NETAPI_GET/SET`. So

> Could you rephrase your proposition for using NETAPI_GET/SET.
If I understand right, you would propose to use it to replace the direct IPC calls, right?

Yes.

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 6 above)
Comment (BytesGalore) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: I migrated the responsibility back to IPv6 [1]. 
The actual forwarding of the packet is now performed when the demultiplexing is finished [2].


[1] https://github.com/RIOT-OS/RIOT/pull/7231/files#diff-74976d45a42eaa063e3561b5fd36f615R146
[2] https://github.com/RIOT-OS/RIOT/pull/7231/files#diff-74976d45a42eaa063e3561b5fd36f615R255

File: sys/include/net/netopt.h [hunk 7]
```diff
@@ -269,6 +269,18 @@ typedef enum {
      */
     NETOPT_LAST_ED_LEVEL,
 
+    /**
+     * @brief   Type for sequiential processing of IPv6 extension headers
+     *
+     * On GET it requests a thread to insert an extension header.
```
Comment (miri64) [inline sys/include/net/netopt.h]: Why does get not just return a `gnrc_pktsnip_t` and the IPv6 thread inserts the header?

Comment (miri64) [inline sys/include/net/netopt.h]: (then it still could be used like the other get/set calls ;-)).

Comment (BytesGalore) [inline sys/include/net/netopt.h]: ok, I gave it a shot

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c [hunk 8]
```diff
@@ -97,6 +97,109 @@ kernel_pid_t gnrc_ipv6_init(void)
...
+
+            if (data) {
+                /* insert the extension header snip */
+                ipv6_hdr_t* hdr = gnrc_ipv6_get_header(pkt);
+                if (hdr && (pkt->data == (void *)hdr)) {
+                    /* data has been placed just below the IPv6 Header,
+                     * i.e. in front of hdr->next.
+                     * We bend hdr->next to point to ext.
+                     */
+                    pkt->next = data;
```
Comment (miri64) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: Don't you need to append the old payload of `pkt` as well?

Comment (miri64) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: Ah got it. The handler already does that for you.

File: sys/include/net/netopt.h [hunk 9]
```diff
@@ -269,6 +269,18 @@ typedef enum {
      */
     NETOPT_LAST_ED_LEVEL,
 
+    /**
+     * @brief   Type for sequiential processing of IPv6 extension headers
+     *
+     * On GET it requests a thread to provide an extension header.
```
Comment (miri64) [inline sys/include/net/netopt.h]: "and appends it to the the given (IPv6) header"

File: sys/include/net/netopt.h [hunk 10]
```diff
@@ -269,6 +269,18 @@ typedef enum {
...
+     * On SET it always returns -ENOTSUP.
+     *
+     * When the header is done the thread replies to the call.
+     * The result is the pktsnip_t with the extension header
+     * The result is NULL if no extension header is passed.
```
Comment (miri64) [inline sys/include/net/netopt.h]: You hand over a pointer (not a pointer of a pointer) below, so this should be impossible. Why not use the numeric return-value of NETAPI_GET (which is returned by `gnrc_netapi_get()`)?, e.g. `sizeof(gnrc_pktsnip_t)` on success, `-ENOMEM` (or whatever is applicable) on error.

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 8 above)
Comment (BytesGalore) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: It depends how the pktsnip is created. 
When we create it with the pkt prepending to pkt->next I would expect that the former content is appended behind the newly created snip.
e.g. `gnrc_pktbuf_add(pkt->next, ....);`

File: sys/include/net/netopt.h (hunk 10 above)
Comment (BytesGalore) [inline sys/include/net/netopt.h]: > not a pointer of a pointer

true, it would require some wrapping.
You mean exploiting the int return type for pointer passing, why not.

File: examples/gnrc_networking/Makefile [hunk 11]
```diff
@@ -34,7 +34,8 @@ USEMODULE += ps
 USEMODULE += netstats_l2
 USEMODULE += netstats_ipv6
 USEMODULE += netstats_rpl
-
+USEMODULE += gnrc_ipv6_ext
+USEMODULE += gnrc_rpl_hop
```
Comment (cgundogan) [inline examples/gnrc_networking/Makefile]: please adjust the dependencies for `gnrc_rpl_hop` to include `gnrc_ipv6_ext` automagically

**Comment by BytesGalore** [comment]
rebased

**Comment by smlng** [comment]
any update in progress here? Otherwise we need to postpone this (again).

**Comment by cgundogan** [comment]
@BytesGalore if you find the time, then this PR needs another rebase. This PR is the first step for non-storing MOP support

**Comment by tcschmidt** [comment]
@BytesGalore rebase date approaching?

File: sys/include/net/gnrc/ipv6.h [hunk 12]
```diff
@@ -145,13 +145,14 @@ kernel_pid_t gnrc_ipv6_init(void);
...
- * @param[in] nh        A protocol number (see @ref net_protnum) of the current snip.
+ * @param[in] netif      The receiving interface.
+ * @param[in] current    A snip to process.
+ * @param[in] pkt        A packet.
+ * @param[in] nh         A protocol number (see @ref net_protnum) of the current snip.
+ * @param[in] is_for_me  Indicates if we are the destination of the packet.
  */
 void gnrc_ipv6_demux(gnrc_netif_t *netif, gnrc_pktsnip_t *current,
-                     gnrc_pktsnip_t *pkt, uint8_t nh);
+                     gnrc_pktsnip_t *pkt, uint8_t nh, bool is_for_me);
```
Comment (miri64) [inline sys/include/net/gnrc/ipv6.h]: I have some refactoring work regarding extension headers laying around (not even pushed yet) that would make this additional parameter most likely unnecessary. This would however throw back this PR again... So I think it makes more sense to merge this first and then clean it up with my refactoring ^^.

Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: There is a lot of weird stuff happening with the extension headers at the moment that can be implemented much more straight forward, that is what the refactoring is for ;-)

File: sys/include/net/gnrc/rpl/hop.h (hunk 1 above)
Comment (miri64) [reply sys/include/net/gnrc/rpl/hop.h]: Alright

File: sys/include/net/ipv6/ext.h [hunk 13]
```diff
@@ -31,6 +33,19 @@ extern "C" {
...
+
+/**
+ * @brief   Message content container to process extension headers sequentially.
+ */
+typedef struct {
+    gnrc_netif_t    *netif;     /**< The interface we received the packet */
+    gnrc_pktsnip_t  *current;   /**< The snip of the IP6 header */
+    gnrc_pktsnip_t  *next_hdr;  /**< The next header to be processed */
+    uint8_t         nh_type;    /**< The type of the next header */
+} gnrc_ipv6_ext_hdr_handle_t;
```
Comment (miri64) [inline sys/include/net/ipv6/ext.h]: I think this however by its very name should go into GNRC not `net/ipv6/ext.h` (which is outside the GNRC realm).

Comment (miri64) [reply sys/include/net/ipv6/ext.h]: (not just the name, also the fact that you are using GNRC-specific types here)

**Comment by miri64** [comment]
See https://github.com/RIOT-OS/RIOT/pull/10244

File: sys/include/net/ipv6/ext.h (hunk 13 above)
Comment (BytesGalore) [reply sys/include/net/ipv6/ext.h]: Don't know, its purpose is only to enable the sequential iteration of extension headers for IPv6 packets.
Although the structure uses gnrc-types I would expect to find such helper in this extension specific headerfile.
Probably I can just remove the prefix `gnrc_` from the typename to avoid confusion.

File: sys/include/net/gnrc/ipv6.h (hunk 12 above)
Comment (BytesGalore) [reply sys/include/net/gnrc/ipv6.h]: It would be great to merge this aged PR finally.
This one (or equivalent) is required to start implementing Non-Stroing MOP for RPL.

File: sys/include/net/ipv6/ext.h (hunk 13 above)
Comment (miri64) [reply sys/include/net/ipv6/ext.h]: @BytesGalore before you do anything, have a look at my refactoring work first of the past few weeks. It also should very much simplify the handling of the hop-by-hop-header, as it is now treated separately (which we can do, because they always MUST come first according to RFC8200; see https://github.com/RIOT-OS/RIOT/pull/10244). You should now be able to just hook into the according switch-branch with `gnrc_ipv6_ext_demux()` for that handling.

File: sys/include/net/gnrc/ipv6.h (hunk 12 above)
Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: I know :-/, but I also finally wanted to clean-up IPv6 reception. I can help you to adapt it to the refactoring!

Comment (miri64) [reply sys/include/net/gnrc/ipv6.h]: (Then someone else of course needs to help reviewing)

Comment (BytesGalore) [reply sys/include/net/gnrc/ipv6.h]: Yeah, I'm convinced that it doesn't make sense to merge this PR now.
There's no use to have this extension header handling in, since now it would be more an obstacle then helpful.
I will wait and refactor it to the coming changes :)

**Comment by miri64** [comment]
Needs rebase / rework, now that #10244 is merged.

**Comment by miri64** [comment]
Ping @BytesGalore?

**Comment by miri64** [comment]
@BytesGalore are you still able to work on this or shall I take it over?

**Comment by smlng** [comment]
ping @BytesGalore, should someone take over?

**Comment by miri64** [comment]
From what I can tell, parts of RFC 6550 are slated to be deprecated by https://datatracker.ietf.org/doc/draft-ietf-roll-useofrplinfo/ (mostly the type field gets a rework). I can take this over, but more on the side, so adapting to current master and the new draft might take a bit.

**Comment by tcschmidt** [comment]
@miri64 don't see any problems taking this over ...

**Comment by miri64** [comment]
> @miri64 don't see any problems taking this over ...

Already in progress. First step would be to merge #13275 ;-).

**Comment by miri64** [comment]
Started migration in https://github.com/miri64/RIOT/tree/gnrc_rpl_opt/feat/initial.

**Comment by tinstructor** [comment]
> From what I can tell, parts of RFC ~6550~ 6553 are slated to be deprecated by https://datatracker.ietf.org/doc/draft-ietf-roll-useofrplinfo/ (mostly the type field gets a rework). I can take this over, but more on the side, so adapting to current master and the new draft might take a bit.

They might be deprecated yes. However, nobody will use the RPI option uncompressed anyway when RIOT finally implements the 6LoRH specified by [RFC 8138](https://tools.ietf.org/html/rfc8138). However, I don't think anybody is working on implementing 6LoRH currently, is there?

**Comment by miri64** [comment]
> However, I don't think anybody is working on implementing 6LoRH currently, is there?

AFAIK, no, but we need the option and support for the paging dispatch first, right?

**Comment by tinstructor** [comment]
@miri64 correct, paging dispatch is needed first

**Comment by mguetschow** [comment]
Since this PR has been stale for almost four years, but still contains useful code and discussion for follow-up, I'll convert it to a draft.

Please feel free to remove the draft state if anyone wants to pick this up again.

**Comment by miri64** [comment]
Maybe could be interesting for @elenaf9?
